File: appliedmathbasics.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.integrate import quad, solve_ivp
from scipy.optimize import minimize
from numpy.linalg import eig, solve
from scipy.fft import fft, ifft

# 1. Linear Algebra: Eigenvalues and Eigenvectors
def eigenvalues_eigenvectors():
    print("1. Linear Algebra: Eigenvalues and Eigenvectors")
    A = np.array([[4, -2], [1, 1]])  # Example matrix
    values, vectors = eig(A)
    print("Matrix A:\n", A)
    print("Eigenvalues:", values)
    print("Eigenvectors:\n", vectors)
    print()

# 2. Numerical Integration
def numerical_integration():
    print("2. Numerical Integration")
    # Integrate f(x) = x^2 from 0 to 1
    def f(x):
        return x ** 2

    result, _ = quad(f, 0, 1)
    print("Integral of x^2 from 0 to 1:", result)
    print()

# 3. Ordinary Differential Equations (ODEs)
def solve_ode():
    print("3. Ordinary Differential Equations (ODEs)")
    # Solve dy/dx = -2y, y(0) = 1
    def dydx(t, y):
        return -2 * y

    t_span = (0, 5)
    y0 = [1]
    sol = solve_ivp(dydx, t_span, y0, t_eval=np.linspace(0, 5, 100))

    # Plot solution
    plt.plot(sol.t, sol.y[0], label="y(t)")
    plt.title("Solution of dy/dx = -2y")
    plt.xlabel("t")
    plt.ylabel("y")
    plt.legend()
    plt.show()

# 4. Optimization
def optimization():
    print("4. Optimization")
    # Minimize f(x) = (x-3)^2
    def f(x):
        return (x - 3) ** 2

    result = minimize(f, x0=0)
    print("Optimal value of x:", result.x[0])
    print("Minimum value of f(x):", result.fun)
    print()

# 5. Fourier Analysis
def fourier_analysis():
    print("5. Fourier Analysis")
    # Generate a sine wave and compute its Fourier transform
    t = np.linspace(0, 1, 500, endpoint=False)
    signal = np.sin(2 * np.pi * 50 * t) + np.sin(2 * np.pi * 120 * t)
    freq = fft(signal)

    # Plot original signal and its frequency spectrum
    plt.figure(figsize=(12, 5))
    plt.subplot(1, 2, 1)
    plt.plot(t, signal)
    plt.title("Original Signal")
    plt.subplot(1, 2, 2)
    plt.plot(np.abs(freq))
    plt.title("Frequency Spectrum")
    plt.show()

# 6. Partial Differential Equations (Heat Equation)
def solve_heat_equation():
    print("6. Partial Differential Equations (Heat Equation)")
    # Solve u_t = alpha*u_xx on [0,1] with u(0,t)=u(1,t)=0
    nx, nt = 50, 100  # Grid points in space and time
    x = np.linspace(0, 1, nx)
    t = np.linspace(0, 0.2, nt)
    alpha = 0.01
    dx, dt = x[1] - x[0], t[1] - t[0]

    u = np.zeros((nx, nt))
    u[:, 0] = np.sin(np.pi * x)  # Initial condition

    for n in range(0, nt - 1):
        for i in range(1, nx - 1):
            u[i, n + 1] = u[i, n] + alpha * dt / dx**2 * (u[i + 1, n] - 2 * u[i, n] + u[i - 1, n])

    plt.imshow(u, extent=[0, 0.2, 0, 1], origin="lower", aspect="auto", cmap="hot")
    plt.colorbar(label="Temperature")
    plt.title("Heat Equation Solution")
    plt.xlabel("Time")
    plt.ylabel("Space")
    plt.show()

# 7. Stochastic Processes (Monte Carlo Simulation)
def monte_carlo_simulation():
    print("7. Stochastic Processes: Monte Carlo Simulation")
    # Estimate pi using random sampling
    np.random.seed(42)
    num_points = 10000
    x = np.random.uniform(-1, 1, num_points)
    y = np.random.uniform(-1, 1, num_points)
    inside_circle = x**2 + y**2 <= 1

    pi_estimate = 4 * np.sum(inside_circle) / num_points
    print("Estimated value of pi:", pi_estimate)

    # Plot points
    plt.figure(figsize=(6, 6))
    plt.scatter(x, y, c=inside_circle, cmap="coolwarm", s=1)
    plt.title("Monte Carlo Simulation of Pi")
    plt.show()

# 8. Regression Analysis (Linear Regression)
def regression_analysis():
    print("8. Regression Analysis")
    # Generate synthetic data
    np.random.seed(42)
    X = np.linspace(0, 10, 100)
    Y = 3 * X + 5 + np.random.normal(0, 2, size=100)

    # Perform linear regression
    A = np.vstack([X, np.ones(len(X))]).T
    coef, _, _, _ = np.linalg.lstsq(A, Y, rcond=None)

    print("Linear regression coefficients (slope, intercept):", coef)

    # Plot
    plt.scatter(X, Y, label="Data")
    plt.plot(X, coef[0] * X + coef[1], color="red", label="Fitted Line")
    plt.title("Linear Regression")
    plt.xlabel("X")
    plt.ylabel("Y")
    plt.legend()
    plt.show()

# Run the demonstrations
eigenvalues_eigenvectors()
numerical_integration()
solve_ode()
optimization()
fourier_analysis()
solve_heat_equation()
monte_carlo_simulation()
regression_analysis()

File: appliedmathbasics.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <random>
#include <algorithm>
#include <functional>
#include <Eigen/Dense> // For linear algebra
#include <gsl/gsl_integration.h> // GNU Scientific Library for integration
#include <matplot/matplot.h> // For plotting (install matplot++ for this to work)

// 1. Linear Algebra: Eigenvalues and Eigenvectors
void eigenvalues_eigenvectors() {
    std::cout << "1. Linear Algebra: Eigenvalues and Eigenvectors\n";
    Eigen::Matrix2d A;
    A << 4, -2, 1, 1;

    Eigen::EigenSolver<Eigen::Matrix2d> solver(A);
    std::cout << "Matrix A:\n" << A << "\n";
    std::cout << "Eigenvalues:\n" << solver.eigenvalues() << "\n";
    std::cout << "Eigenvectors:\n" << solver.eigenvectors() << "\n\n";
}

// 2. Numerical Integration
double f(double x, void* params) {
    return x * x; // Example: Integrand is f(x) = x^2
}

void numerical_integration() {
    std::cout << "2. Numerical Integration\n";
    gsl_integration_workspace* w = gsl_integration_workspace_alloc(1000);
    gsl_function F;
    F.function = &f;

    double result, error;
    gsl_integration_qags(&F, 0, 1, 1e-7, 1e-7, 1000, w, &result, &error);

    std::cout << "Integral of x^2 from 0 to 1: " << result << "\n\n";
    gsl_integration_workspace_free(w);
}

// 3. Ordinary Differential Equations (ODEs)
void solve_ode() {
    std::cout << "3. Ordinary Differential Equations (ODEs)\n";
    // ODE solver logic here (similar to solving dy/dx = -2y)
    // Can use Boost libraries or custom implementation
    std::cout << "ODE solved (not shown, use Boost Odeint for detailed implementation).\n\n";
}

// 4. Optimization
void optimization() {
    std::cout << "4. Optimization\n";
    auto f = [](double x) { return (x - 3) * (x - 3); };
    double x = 0.0, step = 0.01, grad;

    // Gradient descent example
    for (int i = 0; i < 1000; ++i) {
        grad = 2 * (x - 3);
        x -= step * grad;
        if (std::fabs(grad) < 1e-6) break;
    }

    std::cout << "Optimal value of x: " << x << "\nMinimum value of f(x): " << f(x) << "\n\n";
}

// 5. Fourier Analysis
void fourier_analysis() {
    std::cout << "5. Fourier Analysis\n";
    const int N = 500;
    std::vector<double> signal(N), freq(N);

    for (int i = 0; i < N; ++i) {
        signal[i] = sin(2 * M_PI * 50 * i / N) + sin(2 * M_PI * 120 * i / N);
    }

    matplot::plot(signal);
    matplot::title("Original Signal");
    matplot::show();
}

// 6. Partial Differential Equations (Heat Equation)
void solve_heat_equation() {
    std::cout << "6. Partial Differential Equations (Heat Equation)\n";
    const int nx = 50, nt = 100;
    double alpha = 0.01, dx = 1.0 / nx, dt = 0.2 / nt;

    std::vector<std::vector<double>> u(nx, std::vector<double>(nt, 0.0));
    for (int i = 0; i < nx; ++i) u[i][0] = sin(M_PI * i * dx); // Initial condition

    for (int n = 0; n < nt - 1; ++n) {
        for (int i = 1; i < nx - 1; ++i) {
            u[i][n + 1] = u[i][n] + alpha * dt / (dx * dx) * (u[i + 1][n] - 2 * u[i][n] + u[i - 1][n]);
        }
    }

    matplot::imshow(u);
    matplot::colorbar();
    matplot::title("Heat Equation Solution");
    matplot::show();
}

// 7. Stochastic Processes (Monte Carlo Simulation)
void monte_carlo_simulation() {
    std::cout << "7. Stochastic Processes: Monte Carlo Simulation\n";
    const int num_points = 10000;
    std::default_random_engine generator;
    std::uniform_real_distribution<double> distribution(-1.0, 1.0);

    int inside_circle = 0;
    for (int i = 0; i < num_points; ++i) {
        double x = distribution(generator);
        double y = distribution(generator);
        if (x * x + y * y <= 1.0) ++inside_circle;
    }

    double pi_estimate = 4.0 * inside_circle / num_points;
    std::cout << "Estimated value of pi: " << pi_estimate << "\n\n";
}

// 8. Regression Analysis (Linear Regression)
void regression_analysis() {
    std::cout << "8. Regression Analysis\n";
    const int n = 100;
    std::vector<double> X(n), Y(n);
    double slope = 3.0, intercept = 5.0;

    std::default_random_engine generator;
    std::normal_distribution<double> noise(0.0, 2.0);

    for (int i = 0; i < n; ++i) {
        X[i] = i;
        Y[i] = slope * X[i] + intercept + noise(generator);
    }

    double sum_x = std::accumulate(X.begin(), X.end(), 0.0);
    double sum_y = std::accumulate(Y.begin(), Y.end(), 0.0);
    double sum_xy = std::inner_product(X.begin(), X.end(), Y.begin(), 0.0);
    double sum_xx = std::inner_product(X.begin(), X.end(), X.begin(), 0.0);

    double a = (n * sum_xy - sum_x * sum_y) / (n * sum_xx - sum_x * sum_x);
    double b = (sum_y - a * sum_x) / n;

    std::cout << "Linear regression coefficients:\nSlope: " << a << ", Intercept: " << b << "\n";
}

// Main function
int main() {
    eigenvalues_eigenvectors();
    numerical_integration();
    solve_ode();
    optimization();
    fourier_analysis();
    solve_heat_equation();
    monte_carlo_simulation();
    regression_analysis();

    return 0;
}
